cluster/photon-controller/util.sh

```diff
@@ -0,0 +1,995 @@
+    if [ $rc -eq 0 ]; then
+        $PHOTON vm stop $vm_id > /dev/null 2>&1 || rc=$?
+        if [ $rc -ne 0 ]; then
+            kube::log::error "Error: could not stop ${vm_name} ($vm_id)"
+            kube::log::error "Please investigate and stop manually"
+            return
+        fi
+    fi
+
+    rc=0
+    $PHOTON vm delete $vm_id > /dev/null 2>&1 || rc=$?
+    if [ $rc -ne 0 ]; then
+        kube::log::error "Error: could not delete ${vm_name} ($vm_id)"
+        kube::log::error "Please investigate and delete manually"
+    fi
+}
+
+#
+# Looks for the image named PHOTON_IMAGE
+# Sets PHOTON_IMAGE_ID to be the id of that image.
+# We currently assume there is exactly one image with name
+#
+function find-image-id {
+    local rc=0
+    PHOTON_IMAGE_ID=$($PHOTON image list | head -1 | grep $'\t'${PHOTON_IMAGE}$'\t' | grep READY | awk -F'\t' '{print $1}')
+    if [ $rc -ne 0 ]; then
+        kube::log::error "Cannot find image \"${PHOTON_IMAGE}\""
+        fail=1
+    fi
+}
+
+#
+# Generate an ISO with a single file called user-data.txt
+# This ISO will be used to configure cloud-init (which is already
+# on the VM). We will tell cloud-init to create the kube user/group
+# and give ourselves the ability to ssh to the VM with ssh. We also
+# allow people to ssh with the same password that was randomly
+# generated for access to Kubernetes as a backup method.
+#
+# Assumes environment variables:
+#   - VM_USER
+#   - KUBE_PASSWORD (randomly generated password)
+#
+function gen-cloud-init-iso {
+    local password_hash
+    password_hash=$(openssl passwd -1 ${KUBE_PASSWORD})
+
+    local ssh_key
+    ssh_key=$(ssh-add -L | head -1)
+
+    # Make the user-data file that will be used by cloud-init
+    (
+        echo "#cloud-config"
+        echo ""
+        echo "groups:"
+        echo "  - ${VM_USER}"
+        echo ""
+        echo "users:"
+        echo "  - name: ${VM_USER}"
+        echo "    gecos: Kubernetes"
+        echo "    primary-group: ${VM_USER}"
+        echo "    lock-passwd: false"
+        echo "    passwd: ${password_hash}"
+        echo "    ssh-authorized-keys: "
+        echo "      - ${ssh_key}"
+        echo "    sudo: ALL=(ALL) NOPASSWD:ALL"
+        echo "    shell: /bin/bash"
+        echo ""
+        echo "hostname:"
+        echo "  - hostname: kube"
+    ) > ${KUBE_TEMP}/user-data.txt
+
+    # Make the ISO that will contain the user-data
+    # The -rock option means that we'll generate real filenames (long and with case)
+    run-cmd "mkisofs -rock -o ${KUBE_TEMP}/cloud-init.iso ${KUBE_TEMP}/user-data.txt"
+}
+
+#
+# Generate a script used to install salt on the master
+# It is placed into $KUBE_TEMP/master-start.sh
+#
+function gen-master-start {
+    python "${KUBE_ROOT}/third_party/htpasswd/htpasswd.py" \
+        -b -c "${KUBE_TEMP}/htpasswd" "$KUBE_USER" "$KUBE_PASSWORD"
+    local htpasswd
+    htpasswd=$(cat "${KUBE_TEMP}/htpasswd")
+
+    (
+        echo "#! /bin/bash"
```

Comment:
Are you suggesting a here document? I didn't do that because the subshell run commands to generate the file (i.e. grep). I do like the syntax of here documents when applicable, but I don't think it is here.